File: rpi_system/io_controllers/gpio_controller.py
"""
Direct GPIO controller for Raspberry Pi.
Manages HC-SR04 ultrasonic distance sensor wired directly to Pi 40-pin header.
Emulates the 3-second non-blocking cooldown logic from arduino_controller.cpp.
"""
import time
import logging
from typing import Optional

logger = logging.getLogger("EdgeAI.GPIOController")

class RaspberryPiGPIOController:
    """Controls HC-SR04 sensor directly using Raspberry Pi GPIO pins."""

    def __init__(
        self,
        trigger_pin: int = 23,
        echo_pin: int = 24,
        distance_threshold_cm: float = 40.0,
        min_distance_cm: float = 2.0,
        cooldown_seconds: float = 3.0,
    ):
        self.trigger_pin = trigger_pin
        self.echo_pin = echo_pin
        self.distance_threshold_cm = distance_threshold_cm
        self.min_distance_cm = min_distance_cm
        self.cooldown_seconds = cooldown_seconds

        self.last_trigger_time = 0.0
        self.is_ready = True
        self.gpio_available = False
        self._setup_gpio()

    def _setup_gpio(self):
        """Initializes RPi.GPIO or gpiozero if available."""
        try:
            import RPi.GPIO as GPIO
            self.GPIO = GPIO
            self.GPIO.setmode(self.GPIO.BCM)
            self.GPIO.setwarnings(False)
            self.GPIO.setup(self.trigger_pin, self.GPIO.OUT)
            self.GPIO.setup(self.echo_pin, self.GPIO.IN)
            self.GPIO.output(self.trigger_pin, False)
            time.sleep(0.5)
            self.gpio_available = True
            logger.info(
                f"Raspberry Pi GPIO initialized. Trig=BCM{self.trigger_pin}, Echo=BCM{self.echo_pin}"
            )
        except Exception as e:
            logger.warning(
                f"RPi.GPIO not available ({e}). Running in GPIO SIMULATION mode."
            )
            self.gpio_available = False

    def measure_distance_cm(self) -> Optional[float]:
        """
        Triggers an ultrasonic pulse and measures echo duration.
        Returns distance in centimeters, or None if timed out.
        """
        if not self.gpio_available:
            # Return simulated distance for offline testing
            return 25.0

        # Send 10 microsecond trigger pulse
        self.GPIO.output(self.trigger_pin, True)
        time.sleep(0.00001)
        self.GPIO.output(self.trigger_pin, False)

        pulse_start = time.time()
        timeout_start = pulse_start

        # Wait for echo to start
        while self.GPIO.input(self.echo_pin) == 0:
            pulse_start = time.time()
            if pulse_start - timeout_start > 0.04:  # 40ms timeout
                return None

        pulse_end = pulse_start
        timeout_echo = pulse_start

        # Wait for echo to end
        while self.GPIO.input(self.echo_pin) == 1:
            pulse_end = time.time()
            if pulse_end - timeout_echo > 0.04:
                return None

        # Calculate distance
        pulse_duration = pulse_end - pulse_start
        distance = (pulse_duration * 34300) / 2.0
        return round(distance, 1)

    def check_for_trigger(self) -> bool:
        """
        Checks if an object is within the goldilocks zone with non-blocking cooldown.
        Replicates arduino_controller.cpp logic:
          - distance > 2cm and < 40cm
          - cooldown elapsed
        """
        current_time = time.time()

        # Check cooldown
        if (current_time - self.last_trigger_time) < self.cooldown_seconds:
            return False

        if not self.is_ready:
            logger.info("Sensor ready for next trigger.")
            self.is_ready = True

        distance = self.measure_distance_cm()
        if distance is not None:
            if self.min_distance_cm < distance <= self.distance_threshold_cm:
                logger.info(f"Direct GPIO Trigger: Object detected at {distance} cm!")
                self.last_trigger_time = current_time
                self.is_ready = False
                return True

        return False

    def cleanup(self):
        """Cleans up GPIO state."""
        if self.gpio_available:
            try:
                self.GPIO.cleanup()
            except Exception:
                pass

File: rpi_system_C++/src/io_controllers/gpio_controller.cpp
#include "../../include/io_controllers/gpio_controller.hpp"
#include "../../include/config.hpp"
#include <chrono>
#include <spdlog/spdlog.h>
#include <thread>

#ifdef __arm__
#include <pigpio.h>
#endif

GPIOController::GPIOController()
  : trigger_pin(Config::GPIO_TRIGGER_PIN),
    echo_pin(Config::GPIO_ECHO_PIN),
    distance_threshold_cm(Config::TRIGGER_DISTANCE_THRESHOLD_CM),
    min_distance_cm(Config::MIN_DISTANCE_THRESHOLD_CM),
    cooldown_seconds(Config::COOLDOWN_SECONDS),
    is_ready(true),
    gpio_available(false) { setup_gpio(); }
GPIOController::~GPIOController() { cleanup(); }

void GPIOController::setup_gpio() {
#ifdef __arm__
  if (gpioInitialise() < 0) {
    spdlog::warn("pigpio initialization failed, running in sim mode");
    gpio_available = false;
    return;
  }

  // Initialize the input and output pins
  gpioSetMode(trigger_pin, PI_OUTPUT);
  gpioSetMode(echo_pin, PI_INPUT);

  // set Output pin to 0 so it doesnt output a pulse yet
  gpioWrite(trigger_pin, 0);

  // sensor settling time (500ms)
  std::this_thread::sleep_for(std::chrono::milliseconds(500));

  gpio_available = true;
  spdlog::info("Raspberry Pi GPIO initialized. Trig=BCM{}, Echo=BCM{}",
               trigger_pin, echo_pin);

#else
  spdlog::warn(
      "Running on non-Raspberry Pi architecture. GPIO in SIMULATION mode.");
  gpio_available = false;
#endif
}

std::optional<float> GPIOController::measure_distance_cm() {
  if (!gpio_available) {
    return 25.0f;
  }

#ifdef __arm__
  // send a 10 micro second trigger pulse
  gpioWrite(trigger_pin, 1);
  gpioDelay(10); // sleep for 10 micro seconds
  gpioWrite(trigger_pin, 0);

  uint32_t pulse_start = gpioTick();
  uint32_t timeout_start = pulse_start;

  // wait for echo to start
  while (gpioRead(echo_pin) == 0) {
    pulse_start = gpioTick();
    // 40ms timeout (40000 us)
    if ((pulse_start - timeout_start) > 40000) {
      return std::nullopt;
    }
  }
  
  uint32_t pulse_end = pulse_start;
  uint32_t timeout_echo = pulse_start;

  // wait for echo to end
  while (gpioRead(echo_pin) == 1) {
    pulse_end = gpioTick();
    if ((pulse_end - timeout_echo) > 40000) {
      return std::nullopt;
    }
  }

  // calculate distance traveled (speed of sound 34300 cm/s = 0.0343 cm/us)
  float pulse_duration = static_cast<float>(pulse_end - pulse_start);
  float distance = (pulse_duration * 0.0343f) / 2.0f;
  return std::round(distance * 10.0f) / 10.0f;
#else
  return std::nullopt;
#endif
}

bool GPIOController::check_for_trigger() {
  auto now = std::chrono::steady_clock::now();

  // check cooldown
  std::chrono::duration<float> elapsed = now - last_trigger_time;
  if (elapsed.count() < cooldown_seconds) {
    return false;
  }

  // ready message after cooldown
  if (!is_ready) {
    spdlog::info("Sensor ready for next trigger.");
    is_ready = true;
  }

  // measure distance
  std::optional<float> distance = measure_distance_cm();

  // evaluate detection threashold
  if (distance.has_value()) {
    float d = distance.value();
    if ((min_distance_cm <= d) && (d <= distance_threshold_cm)) {
      spdlog::info("Direct GPIO Trigger: Object detected at {} cm!", d);
      last_trigger_time = now;
      is_ready = false;
      return true;
    }
  }
  return false;
}

void GPIOController::cleanup() {
#ifdef __arm__
  if (gpio_available) {
    spdlog::info("Cleaning up GPIO resources");
    // Ensure trigger pin is low
    gpioWrite(trigger_pin, 0);
    // Release hardware resources
    gpioTerminate();
    gpio_available = false;
  }
#else
  gpio_available = false;
#endif
}
